From e1d3ecd9e7192e41f477708b0f2ef2919719bd06 Mon Sep 17 00:00:00 2001
Message-Id: <e1d3ecd9e7192e41f477708b0f2ef2919719bd06.1430272235.git.chang-joon.lee@intel.com>
In-Reply-To: <0282bce11918ec0dd0c4e7e3bd54e392c2b27a51.1430272235.git.chang-joon.lee@intel.com>
References: <0282bce11918ec0dd0c4e7e3bd54e392c2b27a51.1430272235.git.chang-joon.lee@intel.com>
From: Tomas Elf <tomas.elf@intel.com>
Date: Fri, 10 Apr 2015 11:25:46 +0100
Subject: [PATCH 09/10] REVERTME [VPG]: drm/i915: Work around broken TDR hang
 checking by disabling ACTHD check

The Active Head (ACTHD) register behaves erratically under some circumstances
for some reason that is yet to be explained. This causes the TDR hang detection
from working effectively, which sometimes leaves the system in a
irrecoverably hung state when it would otherwise have been recovered
successfully. By disabling the influence of the ACTHD register on the TDR
hang detection algorithm we circumvent this problem until we've got a
better fix.

The ACTHD register is not crucial for effective hang detection and is there
more to reduce the probability of false positives. However, leaving this check
in the hang detection in the state the platform is currently in we are more
likely to get false negatives than we are to get false positives with the check
left out.

We should investigate this strange ACTHD behaviour ASAP so that we can get to
the bottom of the issue and replace this workaround with a more solid fix.

Issue: GMINL-8000
Change-Id: I3382e981cdf4b1d0df9d1ba953d71e73977b3299
Tracked-On: https://jira01.devtools.intel.com/browse/GMINL-8000
Signed-off-by: Tomas Elf <tomas.elf@intel.com>
---
 drivers/gpu/drm/i915/i915_irq.c |   13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/drivers/gpu/drm/i915/i915_irq.c b/drivers/gpu/drm/i915/i915_irq.c
index 2ae0daa..cf8eb51 100644
--- a/drivers/gpu/drm/i915/i915_irq.c
+++ b/drivers/gpu/drm/i915/i915_irq.c
@@ -3590,6 +3590,19 @@ void i915_hangcheck_sample(struct work_struct *work)
 		      (long int) i915_gem_request_get_seqno(last_req),
 		      (idle ? "true" : "false"));
 
+	/*
+	 * ACTHD breaks in some instances by constantly changing even when
+	 * there is no reason for doing so. The underlying reason for this is
+	 * still pending investigation. This behaviour breaks the TDR hang
+	 * detection since it interprets the constantly changing ACTHD as a
+	 * liveliness indication even when there is a hang. To work around this
+	 * problem we disable the ACTHD check until we have investigated why
+	 * this is happening. The hang detection algorithm does not rely on
+	 * ACTHD entirely and will work just as well in the vast majority of
+	 * cases even without checking ACTHD.
+	 */
+	acthd = 0;
+
 	/* Check both head and active head.
 	* Neither is enough on its own - acthd can be pointing within the
 	* batch buffer so is more likely to be moving, but the same
-- 
1.7.9.5

